contracts/MechMarketplace.sol

```diff
@@ -177,30 +175,33 @@ contract MechMarketplace is IErrorsMarketplace {
         );
     }
 
-    /// @dev Calculates payment for request delivery.
+    /// @dev Calculates payment and fee based on delivery rates and mech type.
     /// @param mech Delivery mech address.
-    /// @param payment Payment amount.
+    /// @param deliveryRate Delivery rate.
+    /// @param mechPayment Mech payment.
+    /// @param marketplaceFee Marketplace fee.
     function _calculatePayment(
         address mech,
-        uint256 payment
+        uint256 deliveryRate
     ) internal virtual returns (uint256 mechPayment, uint256 marketplaceFee) {
-        // This must never be zero
-        if (payment > 0) {
-            // Process payment
-            marketplaceFee = (payment * fee) / 10_000;
-            mechPayment = payment - marketplaceFee;
-
-            // Check for zero value
-            if (mechPayment == 0) {
-                revert ZeroValue();
-            }
+        IMech.MechType mechType = IMech(mech).mechType();
+        uint256 maxDeliveryRate = IMech(mech).maxDeliveryRate();
 
-            // Record payment into mech balance
-            mapMechBalances[mech] += mechPayment;
+        // TODO what if fee is zero just because the delivery rate is in the order of 1..10_000?
+        // Calculate mech payment and marketplace fee
+        marketplaceFee = (deliveryRate * fee) / 10_000;
```

Comment:
Not sure.
marketplaceFee = (deliveryRate * fee) / 10_000;
Something is logically wrong here with the dimensions or the name.
marketplaceFee [token?] = Rate [?] * 10 / 10_000
for 
mechPayment = deliveryRate - marketplaceFee; We need the same dimensions for all quantities.

Reply:
Fee is in `10_000` form, but TDB

Reply:
The problem with this approach is the rounding. For micropayments on Nevermined you'll have marketplaceFee round down to 0.

Alternative approach is to simply implement the fee at the withdrawal stage. It's 10x simpler. Simply when a mech withdraws their balance the marketplace takerate gets applied uniformly.